README
IMAProwl Version 1.2

 * Requirements

  - Ruby 1.9.0 or later with OpenSSL module.
  - Prowl service account. (free to get your account, http://prowl.weks.net/)
  - Prowl iPhone application from AppStore, not free.
  - IMAP server should support IMAPs(IMAP over SSL) or TLS.

 * Instruction

  - edit example configuration file and copy it to ~/.imaprowl.conf
  - execute imaprowl.rb with ruby1.9

 * Command line option

  -  -c, --config FILENAME       Specify configuration file.
  -  -q, --daemon                Enable daemon mode.
  -  -d, --debug                 Enable debug mode.

 * Configuration file

  - load path order. (it'll be overridden with -c option)

    1. -c option
    2. $HOME/.imaprowl.conf
    3. `pwd`/imaprowl.conf
    4. `pwd`/config.yml (deprecated)
    5. dirname($0)/imaprowl.conf

  - config description

  LogDir: <String>
   log directory, ex: "/home/foo/logs/".
   If it's not defined, output all logs into STDOUT/ERR and turn on debug mode forcefully.
  Debug: <Boolean>
    [optional] to enable debug output, default: false
  Daemon: <Boolean>
    [optional] to enable daemon mode, default: false

  Prowl:
   APIKey: <String>
    The prowl apikey of your account
   ProxyHost: <String>
   ProxyPort: <Int>
   ProxyUser: <String>
   ProxyPass: <String>
    [optional] HTTP Proxy settings if it's available.

  Accounts:
  -
    Appication: <String>
     application name for prowl. (Identity of account)
    User: <String>
     username for IMAP server.
    Pass: <String>
     password for IMAP server.
    Host: <String>
     hostname or IP address of IMAP server, SSL/TLS support is required.
    Port: <Int>
     [optional] TCP port number of IMAP server, default: 993
    Priority: <Int>
     [optional] Notification Priority: -2(low) to 2(high), default: 0
    MailBox: <String>
     [optional] mailbox to check UNSEEN, default: INBOX
    BodyLength: <Int>
     [optional] body length for prowl notification, default: 100, Max: 1000
    SubjectLength: <Int>
     [optional] subject length for prowl notification, default: 20
    Format: <string>
     [optional] prowl text format, you can use "%{subject}", "%{from}", "%{name}" and "%{addr}" keyword in the format string.  default: "%{subject} (%{from})"
    Timeout: <Int> (in minutes)
     [optional] Timeout for No-Response, default: 20
    NOOPInterval: <Int> (in seconds)
     [optional] NOOP interval for no-IDLE mode, it should be really short about 30 or so. default: 30
    NoIDLE: <Boolean>
     [optional] force no-IDLE mode, default: false.
    Enable: <Boolean>
     [optional] false to disable this account, default: true

-- 
Takuo Kitame.
http://github.com/takuo/


ChangeLog
2009-10-07  Takuo Kitame <kitame@debian.org>

	* imaprowl.rb (login): MailBox name should be UTF-7
	* tagged as 1.2.1

2009-10-06  Takuo Kitame <kitame@debian.org>

	* imaprowl.rb (check_unseen): try body.encode() with src encoding as 'ISO-2022-JP' if body.encoding is US-ASCII. (don't warry ISO-2022-JP is compatible with US-ASCII)

2009-10-05  Takuo Kitame <kitame@debian.org>

	* imaprowl.rb: bump version 1.2.1
	more verbose logging for DEBUG
	(warn): new logger function for Logger::WARN
	(check_unseen): falling back to default format if @format not usable.
	(idler): handle some IMAP responses.

2009-09-29  Takuo Kitame <kitame@debian.org>

  * config: added Format. Interval is now renamed to Timeout
    print/log warning 'Interval is deprecated.'
  * imaprowl.rb (check_unseen): format string can be used for prowl event text.
    simplified string count method.
    (initialize): default value of SubjectLength is 20
  * tagged as 1.2

2009-09-25  Takuo Kitame <kitame@debian.org>

  * imaprowl.rb (check_unseen): always convert body to UTF-8
  * tagged as 1.1.2

2009-09-18  Takuo Kitame <kitame@debian.org>

  * imaprowl.rb: String#encode use option :undef=>:replace and replacement charactor should be default.
  * merged 1.1, tagged as 1.1.1

2009-09-17  Takuo Kitame <kitame@debian.org>

  * imaprowl.rb: don't use Iconv or NKF. use String#encode instead.
  (iconv_mime_decode): renamed to mime_decode()
  (check_unseen): fetch BODY should be BODY.PEEK
  (post_escape): used for escaping string while HTTP/POST
  (prowl): use post_escape() instead of URI#escape
  (check_unseen): remove empty line and white-space of line-head before POST.
  (check_unseen): triming should be done before count length.

2009-09-16  Takuo Kitame <kitame@debian.org>

  * New branch: 1.1
  * imaprowl.rb (get_text_part): try to get TEXT part of a mail.
  (check_unseen): try to get first text part as body if possible.

2009-09-16  Takuo Kitame <kitame@debian.org>

  * imaprowl.rb (check_unseen): should retry to prowl if fail to prowl.
  * debian/*: Depends on ruby1.9.1

  * tagged as 1.0.1

2009-09-15  Takuo Kitame <kitame@debian.org>

  * imaprowl.rb: use Iconv if it's available.
  added Signal.trap(:TERM).
  (iconv_mime_decode): MIME decode and convert to UTF-8.
  (check_unseen): use iconv instead of NKF.
  Don't prowl body if it's contains invalid UTF-8 charactor.
  Subject will be shorten length with value of SubjectLength
  clear @notified cache if no UNSEEN mail.
  rescue all exception while processing text and will prowl with error message.
  nil check before checking variable content.
  (initialize): default SubjectLength is 30
 
  * LICENSE: fixed typo.
  * tagged as 1.0.0

2009-09-10  Takuo Kitame <kitame@debian.org>

  * Added debian/manpage.xml
  * debian/rules (build): added rule for building manpage.
  replace #!/usr/bin/ruby to ruby1.9
  * debian/control (Description): update
  (Build-Depends): added docbook-xsl, xsltproc, docbook-xml
  * README: update
  * config.yml: update
  * Tagged as 0.9.2

2009-09-09  Takuo Kitame <kitame@debian.org>

  * AUTHORS: Added.
  * LINCESE: Ditto.
  * GPL: Ditto.
  * imapidle.rb: Added copyright notice.
  * Tagged as 0.9.1

2009-09-08  takuo  <kitame at debian.org>

  * 0.9
  * imaprowl.rb: STDOUT/STDERR always sync=true
  support daemonize with Process.daemon.
  config file load path order.
  Added some command line options. -c (config file), -q (daemon), -d (debug)

2009-09-07  takuo  <kitame at gmail.com>

  * 0.8.2: fix again.
  * imaprowl.rb (check_unseen): fix error caused by undefined method 'encoding'

2009-09-07  takuo  <kitame at gmail.com>

  * Bug fix release : 0.8.1
  * imaprowl.rb (check_unseen): fix error caused by undefined method 'encoding'
  Catch the exception while parsing a Mail.

2009-09-04  takuo  <kitame at gmail.com>

  * merge non-idle-support branch
  * imaprowl.rb (checker): NOOP interval uses @noop value as seconds
  disabling account if Enable == false	
  * config.yml: update example.
  * README: up-to-date

  * IMAProwl 0.8 -- is now works with not IDLE capable IMAP server.

2009-09-04  takuo  <kitame at gmail.com>

  * imaprowl.rb (@idle_thread): renamed to @loop_thread.
  (@no_idle): boolean flag for non-IDLE supported account.
  (checker): added. noop loop to check new unseen mail.

2009-09-04  takuo  <kitame at gmail.com>

  * imaprowl.rb(check_unseen): should use FETCH BODY.PEEK instead of FETCH BODY.
  * tagged 0.7.1

2009-09-04  takuo  <kitame at gmail.com>

  * imaprowl.rb: Refectoring and Refactoring.
  Hopefully fixed the blocking issue.
  (status): added to check thread availability.
  * up-to-date 0.7

2009-09-01  takuo  <kitame at gmail.com>

  * imaprowl.rb (class): @@logger will be initialize in IMAProwl#new.
  @@conf added to store the global configuration.
  (prowl): refactoring.
  (check_unseen): fetch should call only onece with message set array.
  refactoring.
  change some exception handling.
  refactoring...

2009-08-31  takuo  <kitame at gmail.com>

  * imaprowl.rb (class): prowl_conf and logger are now class variable.
    (debug),(error),(info): Print all log into STDOUT/STDERR if LogDir is undefined.
    (idler): Fix issue that IMAProwl won't reconnect when error occurred at check_unseen().
    Some clean-ups.

2009-08-29  takuo  <kitame at gmail.com>

  * imaprowl.rb: fix does not send priority parameter.

2009-08-28  takuo  <kitame at gmail.com> 

  * README: default interval is 20.
  update, rubygems is no longer required.
  up-to-date 0.6.

  * config.yml: added Priority item in Account

  * imaprowl.rb: no longer use rubygems.
  support HTTP Proxy for prowling.
  configurable Prowl Priority by Account.
  don't output any log when LogDir is undefined.

2009-08-25  takuo  <kitame at gmail.com>

  * .gitignore, README, config.yml, imapidle.rb, imaprowl.rb: 
  0.5 with cleanup Repo


imapidle.rb
#
#
# This code has been copied from Ruby-1.9.2's Net::IMAP.
#
# Copyright (C) 2000  Shugo Maeda <shugo@ruby-lang.org>
#
# This library is distributed under the terms of the Ruby license.
# You can freely distribute/modify this library.
#
#

class Net::IMAP
  def idle(&response_handler)
    raise LocalJumpError, "no block given" unless response_handler

    response = nil

    synchronize do
      tag = Thread.current[:net_imap_tag] = generate_tag
      put_string "#{tag} IDLE#{CRLF}"

      begin
        add_response_handler response_handler

        @idle_done_cond = new_cond
        @idle_done_cond.wait
        @idle_done_cond = nil
      ensure
        remove_response_handler response_handler
        put_string "DONE#{CRLF}"
        response = get_tagged_response( tag, "IDLE")
      end
    end

    response
  end

  def idle_done
    synchronize do
      if @idle_done_cond.nil? 
         raise Net::IMAP::Error, 'not during IDLE'
      end
      @idle_done_cond.signal
    end
  end
end


imaprowl.rb
#!/usr/bin/ruby
#
# IMAProwl - Prowl notification for IMAP new mail
# Version: 1.2
#
# Copyright (c) 2009 Takuo Kitame.
#
# You can redistribute it and/or modify it under the same term as Ruby.
#
STDOUT.sync = STDERR.sync = true

$:.insert(0, File.dirname(__FILE__))

IMAPROWL_VERSION = "1.2.1"
if RUBY_VERSION < "1.9.0"
  STDERR.puts "IMAProwl #{IMAPROWL_VERSION} requires Ruby >= 1.9.0"
  exit
end
$0 = "IMAProwl/#{IMAPROWL_VERSION}"

require 'optparse'
require 'uri'
require 'net/https'
require 'net/imap'
require 'yaml'
require 'logger'
require 'imapidle' unless Net::IMAP.respond_to?("idle")

class IMAProwl

  PROWL_API_ADD = "https://prowl.weks.net/publicapi/add"

  @@conf = Hash.new
  @@logger = nil
  @@prowl_conf = nil

  attr_reader :enable
  attr_reader :loop_thread

  def initialize( global, conf )
    @@conf = global
    _prowl_conf_validate( global['Prowl'] )
    @application = conf['Application'] ? conf['Application'] : "IMAProwl"
    @user = conf['User']
    @pass = conf['Pass']
    @host = conf['Host']
    @port = conf['Port'] ? conf['Port'] : 993
    @mailbox = conf['MailBox'] ? conf['MailBox'] : "INBOX"
    # backward compat
    @timeout = conf['Timeout'] ? conf['Timeout'] : ( conf['Interval'] ? conf['Interval'] : 20 )
    if conf['Interval']
      warn = "Warning: 'Interval' is deprecated. You should use 'Timeout' instead."
      STDERR.print "#{warn}\n"
      warn warn
    end
    @noop = conf['NOOPInterval'] ? conf['NOOPInterval'] : 30
    @subject_length = conf['SubjectLength'] ? conf['SubjectLength'] - 1 : 19
    @body_length = conf['BodyLength'] ? conf['BodyLength'] - 1 : 99
    @body_length = 1 if @body_length < 0
    @subject_length = 1 if @subject_length < 0
    @priority = conf['Priority'] ? conf['Priority'] : 0
    @notified = []
    @enable = conf.has_key?('Enable') ? conf['Enable'] : true
    @no_idle = conf.has_key?('NoIDLE') ? conf['NoIDLE'] : false
	@format = conf.has_key?('Format') ? conf['Format'] : "%{subject} (%{from})"
  end

  # start() should run only once
  def start
    info "Starting."
    connect()
    unless @imap.capability.include?( 'IDLE' )
      error "Error: #{@host} does not support IDLE."
      error "Falling back to no IDLE support mode."
      @no_idle = true
    end
    login()
    check_unseen( false )
    if @no_idle
      checker()
    else
      idler()
    end
  end

  def restart
    info "Restarting..."
    connect()
    login()
    if @no_idle
      checker()
    else
      check_unseen( true )
      idler()
    end
    debug "Restarted"
  end

  def stop
    unless @no_idle
      @idle_time = nil
      @imap.idle_done
      debug "DONE IDLE."
    end
  end

  def status
    return if @no_idle
    retried = false
    debug "Check process status..."
    begin
      if @imap.disconnected?
        @loop_thread.exit if @loop_thread.alive?
        error "socket is disconnected. trying to reconnect..."
        restart
      elsif ! @loop_thread.alive?
        error "IDLE thread is dead."
        restart
      end
      if @timeout > 0 && @idle_time && Time.now - @idle_time > 60 * @timeout
        info "Timeout exceed. "
        stop
      end
    rescue
      @loop_thread.exit if @loop_thread.alive?
      error $!.to_s
      unless retried
        retried = true
        retry
      end
    end
  end

  private
  def post_escape( string )
    string.gsub(/([^ a-zA-Z0-9_.-]+)/) do
      '%' + $1.unpack('H2' * $1.bytesize).join('%').upcase
    end.tr(' ', '+')
  end

  def mime_decode( input, out_charset = 'utf-8' )
    while input.sub!(/(=\?[A-Za-z0-9-]+\?[BQbq]\?[^\?]+\?=)(?:(?:\r\n)?[\s\t])+(=\?[A-Za-z0-9-]+\?[BQbq]\?[^\?]+\?=)/, '\1\2')
    end
    begin
      ret = input.sub!( /=\?([A-Za-z0-9-]+)\?([BQbq])\?([^\?]+)\?=/ ) {
        charset = $1
        enc = $2.upcase
        word = $3
        debug "Decode MIME hedaer: Charset: #{charset}, Encode: #{enc}, Word: #{word}"
        word = word.unpack( { "B"=>"m*", "Q"=>"M*" }[enc] ).first
        # Iconv.conv( out_charset + "//IGNORE", charset, word )
        word.encode( out_charset, charset, :undef=>:replace, :invalid=>:replace )
      }
      return ret ? mime_decode( input ) : input
    rescue
      # "Error while convert MIME string."
      error "Error while converting MIME header: #{input}"
      debug "E: #{$!}"
      return input
    end
  end

  def _prowl_conf_validate( val )
    return if @@prowl_conf
    @@prowl_conf = val
    unless @@prowl_conf.kind_of?( Hash )
      STDERR.print "Configuration Error: Prowl section must be Hash.\n"
      exit 1
    end
    unless @@prowl_conf.has_key?( 'APIKey' )
      STDERR.print "Configuration Error: APIKey must be given.\n"
      exit 1
    end
    _init_logger()
  end

  def _init_logger
    if @@conf['LogDir']
      logdir = @@conf['LogDir']
      Dir.mkdir( logdir ) unless File.exist?( logdir )
      filename = File.join( logdir, "imaprowl.log" )
      STDOUT.puts "All logs will be written into #{filename}."
      @@logger = Logger.new( filename, 'daily' )
      @@logger.level = @@conf['Debug'] ? Logger::DEBUG : Logger::INFO
      @@logger.datetime_format = "%Y-%m-%dT%H:%M:%S"
    else
      @@logger = nil
    end
  end

  def _log( severity, str )
    if @@logger
      @@logger.add( severity, str, @application )
    else
      format = "[%Y-%m-%dT%H:%M:%S##{Process.pid}] #{@application} - #{str}\n"
      if severity == Logger::ERROR
        STDERR.print Time.now.strftime( format )
      else
        print Time.now.strftime( format )
      end
    end
  end

  def debug( str )
    _log( Logger::DEBUG, "#{Thread.current}:#{str}" )
  end

  def error( str )
    _log( Logger::ERROR, "#{Thread.current}:#{str}" )
  end

  def warn( str )
    _log( Logger::WARN, str)
  end

  def info( str )
    _log( Logger::INFO, str )
  end

  def login
    return true if @logged_in
    ret = @imap.login( @user, @pass )
    if ret.name != "OK"
      error "Failed to login: user: #{@user}@#{@host}."
      return false
    end
    @imap.select( Net::IMAP.encode_utf7( @mailbox ) )
    @logged_in = true
    return true
  end

  def connect
    @logged_in = false
    begin
      @imap = Net::IMAP.new( @host, @port, true, nil, false ) # don't verify cert
    rescue
      error "Error on connect()"
    end
  end

  def disconnected?
    return @imap.disconnected?
  end

  def prowl( params = {} )
    uri = URI::parse( PROWL_API_ADD )
    if @@prowl_conf['ProxyHost']
      http = Net::HTTP::Proxy( @@prowl_conf['ProxyHost'],
                               @@prowl_conf['ProxyPort'],
                               @@prowl_conf['ProxyUser'],
                               @@prowl_conf['ProxyPass']).new( uri.host,
                                                               uri.port )
    else
      http = Net::HTTP.new( uri.host, uri.port )
    end

    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_NONE

    request = Net::HTTP::Post.new( uri.request_uri )
    request.content_type = "application/x-www-form-urlencoded"

    query = params.map do |key, val| "#{key}=#{post_escape( val.to_s )}" end

    return http.request(request, query.join('&'))
  end

  def get_text_part( struct, pos )
    if struct.kind_of?( Net::IMAP::BodyTypeMultipart )
      struct.parts.each_index do |i|
        pos.push( i+1 )
        part, pos = get_text_part( struct.parts[i], pos )
        return part, pos if part && part.media_type == "TEXT"
        pos.pop
      end
    end
    if struct.media_type == "TEXT"
      return struct, pos
    end
    return nil, pos
  end

  def check_unseen( will_prowl = false )
    debug "Checking UNSEEN mail."

    unseen = @imap.search( ['UNSEEN'] )
    if unseen.size == 0
      @notified = []
      return
    end

    unseen_set = Array.new

    data_set = @imap.fetch( unseen, "(ENVELOPE BODYSTRUCTURE UID)" )
    data_set.each do |data|
      begin
        attr = data.attr

        if @notified.include?( attr["UID"] )
          debug "SKIP Already notified: UID=#{attr["UID"]}"
          unseen_set.push attr["UID"]
          next
        end

        # header process
        envelope = attr["ENVELOPE"]

        addr = "#{envelope.from.first.mailbox}@#{envelope.from.first.host}"
        name = envelope.from.first.name
        begin
          name = name ? mime_decode( name ) : ""
          from = name != "" ? name : addr
        rescue
          error "Error: Invalid From."
          debug $!.to_s
          from = "[Invalid From]"
        end

        begin
          subject = envelope.subject ? mime_decode( envelope.subject ) : "Untitled"
          if subject.size > @subject_length
            subject = subject[0..@subject_length] + "..."
          end
        rescue
          error "Error: Invalid Subject."
          debug $!.to_s
          subject = "[Invalid Subject]"
        end

        begin
          event = @format % { :subject => subject, :from => from,
            :name=>name, :addr=>addr }
        rescue KeyError
          error "Invalid format string: #{@format}"
          @format = "%{subject} from: %{from}"
          warn "Failing back to default format: #{@format}"
          retry
        rescue ArgumentError
          warn "This ruby does not support format string with Hash argument. falling back to default format."
          event = "#{subject} (#{from})"
        end

        # body process
        begin
          part, pos = get_text_part( attr['BODYSTRUCTURE'], [] )
          if part
            section = pos.size > 0 ? pos.join('.') : "1"
            debug "Detected text part: [#{section}]"
            tmp = @imap.uid_fetch( attr['UID'], "BODY.PEEK[#{section}]" ).first
            body = tmp.attr["BODY[#{section}]"]
          else
            body = "[Body does not contain TEXT part]"
            part = attr['BODYSTRUCTURE']
            debug "No text part found."
          end

          if part.media_type != "TEXT"
            # do nothing
          elsif part.respond_to?('encoding') && part.encoding == "QUOTED-PRINTABLE"
            body = body.unpack("M*").first
          elsif part.respond_to?('encoding') && part.encoding == "BASE64"
            body = body.unpack("m*").first
          end
        
          charset = nil
          if part.param && part.param['CHARSET'] 
            charset = part.param['CHARSET']
          end

          body.force_encoding( "ISO-2022-JP" ) if body.encoding == Encoding::US_ASCII

          debug "Convert body charset from #{charset ? charset : body.encoding.to_s}"
          begin
            body.encode!( "UTF-8", charset, :undef=>:replace, :invalid=>:replace )
          rescue
            error "Error while converting body from #{charset}"
            debug $!.to_s
            body = "[Body contains invalid charactor]"
          end

          body = body.gsub(/^[\s\t]*/, '').gsub(/^$/, '')
          if body.size > @body_length
            body = body[0..@body_length] + "..."
          end
        rescue
          error "Error: Could not parse body text"
          debug $!.to_s
          body = "[Could not parse body]"
        end

        # prowling
        if will_prowl
          info "Prowling... UID=#{attr["UID"]}"
          debug "Prowling: " + event + " " + body
          begin
            presp = prowl( :apikey=> @@prowl_conf['APIKey'],
                           :application => @application,
                           :event => event,
                           :description => body,
                           :priority => @priority
                           )
            unseen_set.push attr["UID"]  if presp && presp.code == "200"
            debug "Response: #{presp.code}"
          rescue
            error "Error while HTTP/POST process."
            debug $!
          end
        else
          unseen_set.push attr["UID"]
          debug "Caching: " + event + " " + body
          debug "Not Prowled (Caching): UID=#{attr["UID"]}"
        end
      rescue
        error "Error while parsing mail: UID=#{attr["UID"]}. Skipped."
        unseen_set.push attr["UID"]
        debug $!
      end
    end
    # caching
    @notified = unseen_set
  end

  def checker
    debug "Won't use IDLE to check unseen mail."
    @loop_thread = Thread.start do
      loop do
        begin
          event = false
          @imap.synchronize do
            @imap.noop
            event = true if @imap.responses["EXISTS"][-1]
            @imap.responses.delete("EXISTS")
          end
          info "Received EXISTS." if event
          check_unseen( true ) if event
          sleep( @noop )
        rescue
          error "Error in checker(): #{$!}"
          debug "Exiting thread"
          Thread.current.exit
        end
      end # loop
    end # Thread
  end

  def idler
    @loop_thread = Thread.start do
      loop do
        begin
          event = false
          debug "Entering IDLE."
          @idle_time = Time.now
          @imap.idle do |resp|
            if resp.kind_of?( Net::IMAP::UntaggedResponse )
              case resp.name
              when "EXISTS"
                event = true
                info "Received EXISTS."
                @idle_time = nil
                @imap.idle_done
                debug "DONE IDLE."
              when "OK"
                info "Received OK" + 
                  resp.data.respond_to('text') ? resp.data.text : ""
              else
                debug "FIXME: Unhandled response: #{resp.name}: #{resp}"
              end
            elsif resp.kind_of?( Net::IMAP::ContinuationRequest )
              debug "Received idling"
            else
              debug "FIXME: Unhandled response: #{resp.name}, #{resp}"
            end
          end
          check_unseen( true ) if event
        rescue
          error "Error in idler(): #{$!}"
          debug "Exiting thread"
          Thread.current.exit
        end
        debug "idler(): Still in loop"
      end # loop
    end # Thread
  end

end # class

## __MAIN__

## command line options
ProgramConfig = Hash.new
opts = OptionParser.new
opts.on( "-c", "--config FILENAME", String, "Specify the config file." ) { |v| ProgramConfig[:config] = v }
opts.on( "-q", "--daemon",nil, "Enable daemon mode.") { |v| ProgramConfig[:daemon] = true }
opts.on( "-d", "--debug", nil, "Enable debug output." ) { |v| ProgramConfig[:debug] = true }
opts.version = IMAPROWL_VERSION
opts.program_name = "imaprowl"
opts.parse!( ARGV )

## config file
config_order = [
  File.join( ENV['HOME'], '.imaprowl.conf' ),
  File.join( Dir.pwd, 'imaprowl.conf' ),
  File.join( Dir.pwd, 'config.yml' ),
  File.join( File.dirname( __FILE__ ), 'imaprowl.conf' )
]

filename = nil
if ProgramConfig[:config]
  if File.exist?( ProgramConfig[:config] )
    filename = ProgramConfig[:config]
  else
    STDERR.print "Configuration file does not exist: #{ProgramConfig[:config]}\n"
    exit 1
  end
else
  config_order.each do |conf|
    next unless File.exist?( conf )
    filename = conf
    break
  end
end
if filename.nil?
  STDERR.print "No configuration file exist.\n"
  STDERR.print "File candidates are:\n"
  STDERR.print config_order.join("\n")
  STDERR.print "\n"
  exit 1
end

STDOUT.print "LoadConf: #{filename}\n" 
config = YAML.load_file( filename )
config["Debug"] = true if ProgramConfig[:debug]


## Daemon mode
if ProgramConfig[:daemon] || config['Daemon']
  begin
    Process.daemon( true, true )
  rescue
    STDERR.print $!
    exit 1
  end
  STDOUT.print "Daemonized. PID=#{Process.pid}\n"
end

## Create Account Thread
application = Array.new
config['Accounts'].each do |account|
  app = IMAProwl.new( config, account )
  next unless app.enable
  app.start()
  application.push( app )
end

## Signal trap
Signal.trap(:INT) {
  application.each do |app|
    app.loop_thread.exit if app.loop_thread.alive?
    app.stop
  end
  sleep 1
  exit
}
Signal.trap(:TERM) {
  application.each do |app|
    app.loop_thread.exit if app.loop_thread.alive?
    app.stop
  end
  sleep 1
  exit
}

## main loop
loop do
  sleep 60
  application.each do |app|
    app.status
  end
end

## __END__
